price-feeder/oracle/provider/okx.go

```diff
@@ -0,0 +1,210 @@
+		Last   string `json:"last"`   // Last traded price ex.: 43508.9
+		Vol24h string `json:"vol24h"` // 24h trading volume ex.: 11159.87127845
+	}
+
+	// OkxTickerResponse defines the response structure of a Okx ticker
+	// request.
+	OkxTickerResponse struct {
+		Data []OkxTickerPair   `json:"data"`
+		Arg  SubscriptionTopic `json:"arg"`
+	}
+
+	SubscriptionTopic struct {
+		Channel string `json:"channel"` // Channel name ex.: tickers
+		InstId  string `json:"instId"`  // Instrument ID ex.: BTC-USDT
+	}
+
+	SubscriptionMsg struct {
+		Op   string              `json:"op"` // Operation ex.: subscribe
+		Args []SubscriptionTopic `json:"args"`
+	}
+)
+
+// NewOkxProvider creates a new OkxProvider
+func NewOkxProvider(ctx context.Context, pairs ...types.CurrencyPair) (*OkxProvider, error) {
+	wsURL := url.URL{
+		Scheme: "wss",
+		Host:   okxHost,
+		Path:   okxPath,
+	}
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(wsURL.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to ws: %+v", err)
+	}
+
+	provider := &OkxProvider{
+		wsURL:            wsURL,
+		wsClient:         wsConn,
+		tickersMap:       &sync.Map{},
+		msReadNewMessage: msReadNewMessage,
+	}
+
+	if err := provider.newTickerSubscription(pairs...); err != nil {
+		return nil, err
+	}
+
+	go provider.handleTickers(ctx)
+
+	return provider, nil
+}
+
+// GetTickerPrices returns the tickerPrices based on the saved map
+func (p OkxProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
+	tickerPrices := make(map[string]TickerPrice, len(pairs))
+
+	for _, currencyPair := range pairs {
+		cp := currencyPair
+		price, err := p.getTickerPrice(cp)
+		if err != nil {
+			return nil, err
+		}
+
+		tickerPrices[cp.String()] = price
+	}
+
+	return tickerPrices, nil
+}
+
+func (p OkxProvider) getTickerPrice(cp types.CurrencyPair) (TickerPrice, error) {
+	instrumentId := getInstrumentId(cp)
+	tickerPair, err := p.getMapTicker(instrumentId)
+	if err != nil {
+		return TickerPrice{}, fmt.Errorf("failed to get %s - %+v", instrumentId, err)
+	}
+
+	return tickerPair.ToTickerPrice()
+}
+
+func (p OkxProvider) getMapTicker(instrumentId string) (OkxTickerPair, error) {
+	value, ok := p.tickersMap.Load(instrumentId)
+	if !ok {
+		return OkxTickerPair{}, fmt.Errorf("ticker not found in map %s", instrumentId)
+	}
+
+	pair, ok := value.(OkxTickerPair)
+	if !ok {
+		return OkxTickerPair{}, fmt.Errorf("ticker found %s, but failed on casting to OkxTickerPair", instrumentId)
+	}
+
+	return pair, nil
+}
+
+func (p OkxProvider) handleTickers(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Duration(p.msReadNewMessage) * time.Millisecond):
+			// time after to avoid asking for prices too frequently
+			messageType, bz, err := p.wsClient.ReadMessage()
+			// if some error occurs continue to try to read the next message
+			if err != nil {
+				continue
```

Comment:
Why would this error sometimes? Does this mean we'd miss prices for certain pairs during a tick?
```suggestion
			if err != nil {
				// if some error occurs continue to try to read the next message
				continue
```